sdks/python/apache_beam/io/datastore/v1/datastoreio.py

```diff
@@ -0,0 +1,289 @@
+
+      return disp_data
+
+  class ReadFn(DoFn):
+    """A DoFn that reads entities from Cloud Datastore, for a given query."""
+
+    def __init__(self, project, namespace=None):
+      super(ReadFromDatastore.ReadFn, self).__init__()
+      self._project = project
+      # using _namespace conflicts with DisplayData._namespace
+      self._datastore_namespace = namespace
+      self._datastore = None
+
+    def start_bundle(self, context):
+      self._datastore = helper.get_datastore(self._project)
+
+    def process(self, p_context, *args, **kwargs):
+      query = p_context.element
+      # Returns an iterator of entities that reads in batches.
+      entities = helper.fetch_entities(self._project, self._datastore_namespace,
+                                       query, self._datastore)
+      return entities
+
+    def display_data(self):
+      disp_data = {'project': self._project}
+
+      if self._datastore_namespace is not None:
+        disp_data['namespace'] = self._datastore_namespace
+
+      return disp_data
+
+  @staticmethod
+  def query_latest_statistics_timestamp(project, namespace, datastore):
+    """Fetches the latest timestamp of statistics from Cloud Datastore.
+
+    Cloud Datastore system tables with statistics are periodically updated.
+    This method fethes the latest timestamp (in microseconds) of statistics
+    update using the `__Stat_Total__` table.
+    """
+    query = helper.make_latest_timestamp_query(namespace)
+    req = helper.make_request(project, namespace, query)
+    resp = datastore.run_query(req)
+    if len(resp.batch.entity_results) == 0:
+      raise RuntimeError("Datastore total statistics unavailable.")
+
+    entity = resp.batch.entity_results[0].entity
+    return datastore_helper.micros_from_timestamp(
+        entity.properties['timestamp'].timestamp_value)
+
+  @staticmethod
+  def get_estimated_size_bytes(project, namespace, query, datastore):
+    """Get the estimated size of the data returned by the given query.
+
+    Cloud Datastore provides no way to get a good estimate of how large the
+    result of a query entity kind being queried, using the __Stat_Kind__ system
+    table, assuming exactly 1 kind is specified in the query.
+    See https://cloud.google.com/datastore/docs/concepts/stats.
+    """
+    kind = query.kind[0].name
+    latest_timestamp = ReadFromDatastore.query_latest_statistics_timestamp(
+        project, namespace, datastore)
+    logging.info('Latest stats timestamp for kind %s is %s',
+                 kind, latest_timestamp)
+
+    kind_stats_query = (
+        helper.make_kind_stats_query(namespace, kind, latest_timestamp))
+
+    req = helper.make_request(project, namespace, kind_stats_query)
+    resp = datastore.run_query(req)
+    if len(resp.batch.entity_results) == 0:
+      raise RuntimeError("Datastore statistics for kind %s unavailable" % kind)
+
+    entity = resp.batch.entity_results[0].entity
```

Comment:
I don't think it is necessary. In this particular case it could be considered harmless, as we are only trying to estimate an approximate size. But it would be a datastore bug if it is sending more entities than what we asked for, which is breaking the expected contract. If we start validating such stuff, then I guess there are many things we need to validate.